base/abstractarray.jl

```diff
@@ -113,10 +113,12 @@ linearindexing{T<:AbstractArray}(::Type{T}) = LinearSlow()
 linearindexing{T<:Array}(::Type{T}) = LinearFast()
 linearindexing{T<:Range}(::Type{T}) = LinearFast()
 
-*(::LinearFast, ::LinearFast) = LinearFast()
-*(::LinearSlow, ::LinearFast) = LinearSlow()
-*(::LinearFast, ::LinearSlow) = LinearSlow()
-*(::LinearSlow, ::LinearSlow) = LinearSlow()
+linearindexing(A::AbstractArray, B::AbstractArray) = linearindexing(linearindexing(A), linearindexing(B))
+linearindexing(A::AbstractArray, B::AbstractArray...) = linearindexing(linearindexing(A), linearindexing(B...))
+linearindexing(::LinearFast, ::LinearFast) = LinearFast()
```

Comment:
To clarify, `LinearFast()` means "linear indexing is fast." When that's not true, we use `CartesianIndex`, which is also fast :smile:. Linear indexing is every-so-slightly (well, maybe a little more than slightly) faster than cartesian indexing for certain array types (most observable with high-dimensional arrays), but it's **catastrophically** slower for others. So all this logic is simply to use linear indexing when we know it's safe (`LinearFast`), but use cartesian indexing (`LinearSlow`) whenever we're unsure or know that linear would be a bad idea.

All this code we're discussing is only about defining and reading traits; once the iteration scheme has been selected, it has no further impact on the cost of indexing.